File: shifai-ios/ShifAI/App/ShifAIError.swift
import Foundation

/// Unified error types for the ShifAI app.
/// Used across all layers (network, database, domain).
enum ShifAIError: Error, LocalizedError, Equatable {

    // ─── Network ───
    case networkUnavailable
    case serverError(statusCode: Int)
    case unauthorized
    case conflict
    case timeout

    // ─── Database ───
    case databaseCorrupted
    case migrationFailed(version: Int)
    case recordNotFound(table: String, id: String)

    // ─── Domain ───
    case insufficientData(required: Int, actual: Int)
    case invalidInput(field: String, reason: String)
    case mlModelUnavailable
    case encryptionFailed

    // ─── Sync ───
    case syncConflict(localDate: Date, remoteDate: Date)
    case syncTimeout
    case mergeFailure

    // ─── Export ───
    case exportTooLarge(sizeMB: Int, maxMB: Int)
    case templateNotFound(name: String)
    case pdfGenerationFailed

    // ─── Auth ───
    case biometricNotAvailable
    case biometricFailed
    case sessionExpired

    var errorDescription: String? {
        switch self {
        case .networkUnavailable: return "Pas de connexion internet"
        case .serverError(let code): return "Erreur serveur (\(code))"
        case .unauthorized: return "Session expirée"
        case .conflict: return "Conflit de données"
        case .timeout: return "Délai dépassé"
        case .databaseCorrupted: return "Base de données corrompue"
        case .migrationFailed(let v): return "Échec de migration v\(v)"
        case .recordNotFound(let t, let id): return "Enregistrement \(id) non trouvé dans \(t)"
        case .insufficientData(let req, let act): return "Données insuffisantes (\(act)/\(req) requis)"
        case .invalidInput(let f, let r): return "Entrée invalide: \(f) — \(r)"
        case .mlModelUnavailable: return "Modèle ML non disponible"
        case .encryptionFailed: return "Échec du chiffrement"
        case .syncConflict: return "Conflit de synchronisation"
        case .syncTimeout: return "Synchronisation expirée"
        case .mergeFailure: return "Échec de fusion"
        case .exportTooLarge(let s, let m): return "Export trop volumineux (\(s)MB, max \(m)MB)"
        case .templateNotFound(let n): return "Template '\(n)' non trouvé"
        case .pdfGenerationFailed: return "Échec génération PDF"
        case .biometricNotAvailable: return "Biométrie non disponible"
        case .biometricFailed: return "Authentification biométrique échouée"
        case .sessionExpired: return "Session expirée"
        }
    }

    /// User-facing recovery suggestion
    var recoverySuggestion: String? {
        switch self {
        case .networkUnavailable: return "Vérifie ta connexion et réessaie."
        case .unauthorized, .sessionExpired: return "Reconnecte-toi à ton compte."
        case .timeout, .syncTimeout: return "Réessaie dans quelques instants."
        case .databaseCorrupted: return "Contacte le support."
        case .insufficientData: return "Continue à logger tes données."
        case .biometricNotAvailable: return "Active Face ID dans les réglages."
        default: return "Réessaie ou contacte le support."
        }
    }
}

File: shifai-android/app/src/main/java/com/shifai/app/ShifAIError.kt
package com.shifai.app

/**
 * Unified error types for the ShifAI app.
 * Used across all layers (network, database, domain).
 */
sealed class ShifAIError(
    val code: String,
    override val message: String,
    val recoverySuggestion: String
) : Exception(message) {

    // ─── Network ───
    class NetworkUnavailable : ShifAIError(
        "NET_UNAVAILABLE", "Pas de connexion internet",
        "Vérifie ta connexion et réessaie."
    )
    class ServerError(statusCode: Int) : ShifAIError(
        "NET_SERVER", "Erreur serveur ($statusCode)",
        "Réessaie dans quelques instants."
    )
    class Unauthorized : ShifAIError(
        "NET_UNAUTHORIZED", "Session expirée",
        "Reconnecte-toi à ton compte."
    )
    class Timeout : ShifAIError(
        "NET_TIMEOUT", "Délai dépassé",
        "Réessaie dans quelques instants."
    )

    // ─── Database ───
    class DatabaseCorrupted : ShifAIError(
        "DB_CORRUPTED", "Base de données corrompue",
        "Contacte le support."
    )
    class MigrationFailed(version: Int) : ShifAIError(
        "DB_MIGRATION", "Échec de migration v$version",
        "Contacte le support."
    )
    class RecordNotFound(table: String, id: String) : ShifAIError(
        "DB_NOT_FOUND", "Enregistrement $id non trouvé dans $table",
        "Réessaie ou contacte le support."
    )

    // ─── Domain ───
    class InsufficientData(required: Int, actual: Int) : ShifAIError(
        "DOM_INSUFFICIENT", "Données insuffisantes ($actual/$required requis)",
        "Continue à logger tes données."
    )
    class InvalidInput(field: String, reason: String) : ShifAIError(
        "DOM_INVALID", "Entrée invalide: $field — $reason",
        "Vérifie tes données et réessaie."
    )
    class MLModelUnavailable : ShifAIError(
        "DOM_ML", "Modèle ML non disponible",
        "Le mode règles sera utilisé."
    )
    class EncryptionFailed : ShifAIError(
        "DOM_ENCRYPT", "Échec du chiffrement",
        "Contacte le support."
    )

    // ─── Sync ───
    class SyncConflict : ShifAIError(
        "SYNC_CONFLICT", "Conflit de synchronisation",
        "Choisis quelle version garder."
    )
    class SyncTimeout : ShifAIError(
        "SYNC_TIMEOUT", "Synchronisation expirée",
        "Réessaie dans quelques instants."
    )

    // ─── Export ───
    class ExportTooLarge(sizeMB: Int, maxMB: Int) : ShifAIError(
        "EXPORT_SIZE", "Export trop volumineux (${sizeMB}MB, max ${maxMB}MB)",
        "Réduis la période sélectionnée."
    )
    class PDFGenerationFailed : ShifAIError(
        "EXPORT_PDF", "Échec génération PDF",
        "Réessaie ou contacte le support."
    )

    // ─── Auth ───
    class BiometricNotAvailable : ShifAIError(
        "AUTH_BIO_NA", "Biométrie non disponible",
        "Active Face ID dans les réglages."
    )
    class BiometricFailed : ShifAIError(
        "AUTH_BIO_FAIL", "Authentification biométrique échouée",
        "Réessaie."
    )
    class SessionExpired : ShifAIError(
        "AUTH_EXPIRED", "Session expirée",
        "Reconnecte-toi à ton compte."
    )
}
